pkg/sql/sem/builtins/generator_builtins.go

```diff
@@ -1891,18 +1899,62 @@ func (*checkConsistencyGenerator) ResolvedType() *types.T {
 
 // Start is part of the tree.ValueGenerator interface.
 func (c *checkConsistencyGenerator) Start(ctx context.Context, _ *kv.Txn) error {
-	resp, err := c.consistencyChecker.CheckConsistency(ctx, c.from, c.to, c.mode)
+	span := roachpb.Span{Key: c.from, EndKey: c.to}
+	// NB: should use IterateRangeDescriptors here which is in the 'upgrade' package
+	// to avoid pulling all into memory. That needs a refactor, though.
+	// kvprober also has some code to iterate in batches.
+	descs, err := kvclient.ScanMetaKVs(ctx, c.txn, span)
 	if err != nil {
 		return err
 	}
-	c.remainingRows = resp.Result
+	for _, v := range descs {
+		var desc roachpb.RangeDescriptor
+		if err := v.ValueProto(&desc); err != nil {
+			return err
+		}
+		if len(desc.StartKey) == 0 {
+			desc.StartKey = keys.MustAddr(keys.LocalMax)
+			// Elide potential second copy we might be getting for r1
+			// if meta1 and meta2 haven't split.
+			// This too should no longer be necessary with IterateRangeDescriptors.
+			if len(c.descs) == 1 {
+				continue
+			}
+		}
+		c.descs = append(c.descs, desc)
+	}
 	return nil
 }
 
 // Next is part of the tree.ValueGenerator interface.
-func (c *checkConsistencyGenerator) Next(_ context.Context) (bool, error) {
+func (c *checkConsistencyGenerator) Next(ctx context.Context) (bool, error) {
 	if len(c.remainingRows) == 0 {
-		return false, nil
+		if len(c.descs) == 0 {
+			return false, nil
+		}
+		// NB: peeling off the spans one by one allows this generator to produce
+		// rows in a streaming manner. If we called CheckConsistency(c.from, c.to)
+		// we would only get the result once all checks have completed and it will
+		// generally be a lot more brittle since an error will completely wipe out
+		// the result set.
+		desc := c.descs[0]
+		c.descs = c.descs[1:]
+		resp, err := c.consistencyChecker.CheckConsistency(
```

Comment:
How does this request behave if the start and end key get out of sync with the range descriptor?

Also, originally, how was this request decomposed into per-range requests? Is it by some magic of `DistSender`?

Next to the magic that previously decomposed this request, is there a corresponding magic that streams the response instead of blocking on it fully? Can we push this behaviour as low as that level, so that we don't need to manually deal with meta ranges up here?